src/apify/scrapy/middlewares/apify_proxy.py

```diff
@@ -0,0 +1,156 @@
+            raise ValueError('Username and password must be provided in the proxy URL.')
+
+        request.meta['proxy'] = url.geturl()
+        basic_auth_header = self._get_basic_auth_header(url.username, url.password)
+        request.headers[b'Proxy-Authorization'] = basic_auth_header
+
+        Actor.log.debug(f'ApifyHttpProxyMiddleware.process_request: updated request.meta={request.meta}')
+
+    def process_exception(
+        self: ApifyHttpProxyMiddleware,
+        request: Request,
+        exception: Exception,
+        spider: Spider,
+    ) -> None | Request:
+        """Process an exception that occurs during request processing.
+
+        Args:
+            request: Scrapy Request object.
+            exception: Exception object.
+            spider: Scrapy Spider object.
+
+        Returns:
+            If a TunnelError occurs, return the request object to halt its processing in the middleware pipeline.
+            Return None otherwise to allow the continuation of request processing.
+        """
+        Actor.log.debug(
+            f'ApifyHttpProxyMiddleware.process_exception: request={request}, exception={exception}, spider={spider}',
+        )
+
+        if isinstance(exception, TunnelError):
+            Actor.log.warning(f'ApifyHttpProxyMiddleware: TunnelError occurred for request="{request}", reason="{exception}", skipping...')
+            return request
+
+        return None
+
+    async def _get_new_proxy_url(self: ApifyHttpProxyMiddleware) -> ParseResult:
+        """Get a new proxy URL.
+
+        Raises:
+            NotConfigured: If creation of the proxy configuration fails.
+
+        Returns:
+            ParseResult: New proxy URL.
+        """
+        # Get proxy configuration, creating it if necessary
+        proxy_cfg = (
+            self._proxy_cfg_internal
+            if isinstance(self._proxy_cfg_internal, ProxyConfiguration)
+            else await Actor.create_proxy_configuration(actor_proxy_input=self._proxy_settings)
+        )
+
+        # If the proxy configuration is still not available, raise an error. However, this should not happen due
+        # to the checks in the `from_crawler` method.
+        if proxy_cfg is None:
+            Actor.log.error('Creation of proxy configuration failed. Check the field "proxyConfiguration" in the Actor input.')
+            raise NotConfigured
+
+        # Store the proxy configuration for future use
+        self._proxy_cfg_internal = proxy_cfg
+
+        # Get a new proxy URL and return it
+        new_url = await proxy_cfg.new_url()
+        return urlparse(new_url)
+
+    def _get_basic_auth_header(self: ApifyHttpProxyMiddleware, username: str, password: str) -> bytes:
+        """Generate a basic authentication header for the given username and password."""
+        string = f'{unquote(username)}:{unquote(password)}'
+        user_pass = to_bytes(string, encoding=self._auth_encoding)
+        return b'Basic ' + b64encode(user_pass)
```

Comment:
I believe this could be a "global" helper function.

Reply:
Sure, it's moved to the `scrapy.utils`.